File: minegame-server/src/main/java/top/nipuru/minegame/server/processor/PlayerDataTransferBukkitProcessor.java
package top.nipuru.minegame.server.processor;

import com.alipay.remoting.AsyncContext;
import com.alipay.remoting.BizContext;
import com.alipay.remoting.rpc.protocol.AsyncUserProcessor;
import lombok.extern.slf4j.Slf4j;
import top.nipuru.minegame.common.message.PlayerDataTransferRequest;
import top.nipuru.minegame.server.BukkitPlugin;
import top.nipuru.minegame.server.game.player.GamePlayer;
import top.nipuru.minegame.server.game.player.DataInfo;
import top.nipuru.minegame.server.util.NmsMethods;

import java.util.HashMap;

@Slf4j
public class PlayerDataTransferBukkitProcessor extends AsyncUserProcessor<PlayerDataTransferRequest> {

    private final BukkitPlugin plugin;

    public PlayerDataTransferBukkitProcessor(BukkitPlugin plugin) {
        this.plugin = plugin;
    }

    @Override
    public void handleRequest(BizContext bizCtx, AsyncContext asyncCtx, PlayerDataTransferRequest request) throws Exception {
        plugin.getServer().getScheduler().runTask(plugin, () -> {
            GamePlayer player = plugin.getPlayer(request.getUniqueId());
            if (player == null) {
                asyncCtx.sendResponse(null);
                return;
            }

            NmsMethods.freezePlayer(player.getBukkitPlayer());         // 冻结玩家 不处理客户端发包
            NmsMethods.removeFromPlayerList(player.getBukkitPlayer()); // 强制移出玩家列表 触发 PlayerQuitEvent 并完成数据的保存

            DataInfo dataInfo = new DataInfo(player.getPlayerId(), player.getDbId(), new HashMap<>());
            try {
                player.pack(dataInfo);
            } catch (Exception e) {
                log.error(e.getMessage(), e);
                asyncCtx.sendException(e);
                return;
            }

            PlayerDataTransferRequest.PlayerDataMessage response = new PlayerDataTransferRequest.PlayerDataMessage()
                    .setPlayerId(dataInfo.getPlayerId())
                    .setDbId(dataInfo.getDbId())
                    .setData(dataInfo.getTables());

            asyncCtx.sendResponse(response);
        });
    }

    @Override
    public String interest() {
        return PlayerDataTransferRequest.class.getName();
    }
}

File: minegame-game/src/main/kotlin/top/nipuru/minegame/game/processor/PlayerDataTransferBukkitProcessor.kt
package top.nipuru.minegame.game.processor

import com.alipay.remoting.AsyncContext
import com.alipay.remoting.BizContext
import com.alipay.remoting.rpc.protocol.AsyncUserProcessor
import org.bukkit.Bukkit
import top.nipuru.minegame.common.message.PlayerDataMessage
import top.nipuru.minegame.common.message.PlayerDataTransferRequest
import top.nipuru.minegame.game.gameplay.player.DataInfo
import top.nipuru.minegame.game.gameplay.player.GamePlayer
import top.nipuru.minegame.game.logger
import top.nipuru.minegame.game.nms.freeze
import top.nipuru.minegame.game.nms.quit
import top.nipuru.minegame.game.playerList
import top.nipuru.minegame.game.plugin

class PlayerDataTransferBukkitProcessor : AsyncUserProcessor<PlayerDataTransferRequest>() {
    @Throws(Exception::class)
    override fun handleRequest(bizCtx: BizContext, asyncCtx: AsyncContext, request: PlayerDataTransferRequest) {
        Bukkit.getScheduler().runTask(plugin, Runnable {
            val player: GamePlayer = playerList.getPlayer(request.uniqueId)

            player.bukkitPlayer.freeze() // 冻结玩家 不处理客户端发包
            player.bukkitPlayer.quit()   // 强制移出玩家列表 触发 PlayerQuitEvent 并完成数据的保存

            val dataInfo = DataInfo(HashMap())

            try {
                player.pack(dataInfo)
            } catch (e: Exception) {
                logger.error(e.message, e)
                asyncCtx.sendException(e)
                return@Runnable
            }

            val response = PlayerDataMessage(player.playerId, player.dbId, dataInfo.tables)
            asyncCtx.sendResponse(response)
        })
    }

    override fun interest(): String {
        return PlayerDataTransferRequest::class.java.name
    }
}
